client/lxd_instances.go

```diff
@@ -211,6 +212,169 @@ func (r *ProtocolLXD) UpdateInstances(state api.InstancesPut, ETag string) (Oper
 	return op, nil
 }
 
+func (r *ProtocolLXD) rebuildInstance(instance api.InstancesPost) (Operation, error) {
+	path, _, err := r.instanceTypeToPath(instance.Type)
+	if err != nil {
+		return nil, err
+	}
+
+	// Send the request
+	op, _, err := r.queryOperation("POST", fmt.Sprintf("%s/%s/rebuild?project=%s", path, url.PathEscape(instance.Name), r.project), instance, "")
+	if err != nil {
+		return nil, err
+	}
+
+	return op, nil
+}
+
+func (r *ProtocolLXD) tryRebuildInstance(req api.InstancesPost, urls []string, op Operation) (RemoteOperation, error) {
+	reflReq := reflect.ValueOf(&req).Elem()
+	var s reflect.Value
+	if !reflect.DeepEqual(req.NewSource, api.InstanceSource{}) {
+		s = reflReq.FieldByName("NewSource")
+	} else {
+		s = reflReq.FieldByName("Source")
+	}
+
+	if len(urls) == 0 {
+		return nil, fmt.Errorf("The source server isn't listening on the network")
+	}
+
+	rop := remoteOperation{
+		chDone: make(chan bool),
+	}
+
+	operation := req.Source.Operation
+
+	// Forward targetOp to remote op
+	go func() {
+		success := false
+		var errors []remoteOperationResult
+		for _, serverURL := range urls {
+			if operation == "" {
+				s.FieldByName("Server").SetString(serverURL)
+			} else {
+				s.FieldByName("Operation").SetString(fmt.Sprintf("%s/1.0/operations/%s", serverURL, url.PathEscape(operation)))
+			}
+
+			op, err := r.rebuildInstance(req)
+			if err != nil {
+				errors = append(errors, remoteOperationResult{URL: serverURL, Error: err})
+				continue
+			}
+
+			rop.handlerLock.Lock()
+			rop.targetOp = op
+			rop.handlerLock.Unlock()
+
+			for _, handler := range rop.handlers {
+				_, _ = rop.targetOp.AddHandler(handler)
+			}
+
+			err = rop.targetOp.Wait()
+			if err != nil {
+				errors = append(errors, remoteOperationResult{URL: serverURL, Error: err})
+
+				if shared.IsConnectionError(err) {
+					continue
+				}
+
+				break
+			}
+
+			success = true
+			break
+		}
+
+		if !success {
+			rop.err = remoteOperationError("Failed instance rebuild", errors)
+			if op != nil {
+				_ = op.Cancel()
+			}
+		}
+
+		close(rop.chDone)
+	}()
+
+	return &rop, nil
+}
+
+// RebuildInstanceFromImage rebuilds an instance from an image.
+func (r *ProtocolLXD) RebuildInstanceFromImage(source ImageServer, image api.Image, req api.InstancesPost) (RemoteOperation, error) {
+	reflReq := reflect.ValueOf(&req).Elem()
+	var s reflect.Value
```

Comment:
This looks like unnecessary duplication of logic and as mentioned above unnecessary complexity.